File: python/ValidParentheses.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s)==0:
            return True
        if len(s)%2 !=0:
            return False
        d = {'(':')','{':'}','[':']'}
        l = []
        for i in s:
            if i in d.keys():
                l.append(i)
            else:
                if len(l)==0 or d[l[-1]] != i:
                    return False
                else:
                    l.pop()
        if len(l)==0:
            return True
        else:
            return False

File: cpp/ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s.length()==0)
            return true;
        if (s.length()%2==1)
            return false;
        stack<char> symbols;
        for(int i = 0; i < s.size(); i++){
            if(s[i] == '(' || s[i] == '[' || s[i] == '{')
                symbols.push(s[i]);
            else{
                if(symbols.size() == 0)
                    return false;

                char match;
                if(s[i] == ')')
                    match = '(';
                else if(s[i] == ']')
                    match = '[';
                else if(s[i] == '}')
                    match = '{';
                char c = symbols.top();
                symbols.pop();

                if(c != match)
                    return false;
            }
        }
        if(symbols.size() != 0)
            return false;
        return true;
    }
};
